src/chart_types/partition_chart/layout/viewmodel/fill_text_layout.ts

```diff
@@ -299,189 +307,254 @@ type GetShapeRowGeometry<C> = (
   padding: Padding,
 ) => RowSpace;
 
-function fill(
-  config: Config,
-  layers: Layer[],
-  allFontSizes: string | any[],
-  measure: TextMeasure,
-  rawTextGetter: RawTextGetter,
-  valueGetter: ValueGetterFunction,
-  formatter: ValueFormatter,
-  textFillOrigins: any[],
-  shapeConstructor: (n: ShapeTreeNode) => any,
-  getShapeRowGeometry: GetShapeRowGeometry<RectangleConstruction> | GetShapeRowGeometry<RingSector>,
-  getRotation: Function,
-  leftAlign: boolean,
-  middleAlign: boolean,
+type ShapeConstructor<C> = (n: ShapeTreeNode) => C;
+
+type NodeWithOrigin = { node: QuadViewModel; origin: PointTuple };
+
+function fill<C>(
+  shapeConstructor: ShapeConstructor<C>,
+  getShapeRowGeometry: GetShapeRowGeometry<C>,
+  getRotation: GetRotation,
 ) {
-  return (node: QuadViewModel, index: number) => {
-    const { maxRowCount, fillLabel } = config;
-
-    const layer = layers[node.depth - 1] || {};
-    const verticalAlignment = middleAlign
-      ? VerticalAlignments.middle
-      : node.depth < layers.length
-      ? VerticalAlignments.bottom
-      : VerticalAlignments.top;
-    const fontSizes = allFontSizes[Math.min(node.depth, allFontSizes.length) - 1];
-    const {
-      textColor,
-      textInvertible,
-      fontStyle,
-      fontVariant,
-      fontFamily,
-      fontWeight,
-      valueFormatter,
-      padding,
-    } = Object.assign(
-      { fontFamily: config.fontFamily, fontWeight: 'normal', padding: 2 },
-      fillLabel,
-      { valueFormatter: formatter },
-      layer.fillLabel,
-      layer.shape,
-    );
+  return function(
+    config: Config,
+    layers: Layer[],
+    measure: TextMeasure,
+    rawTextGetter: RawTextGetter,
+    valueGetter: ValueGetterFunction,
+    formatter: ValueFormatter,
+    leftAlign: boolean,
+    middleAlign: boolean,
+  ) {
+    const { maxRowCount, fillLabel, fontFamily: configFontFamily } = config;
+    return (allFontSizes: Pixels[][], textFillOrigin: PointTuple, node: QuadViewModel): RowSet => {
+      const container = shapeConstructor(node);
+      const rotation = getRotation(node);
+
+      const layer = layers[node.depth - 1] || {};
+      const verticalAlignment = middleAlign
+        ? VerticalAlignments.middle
+        : node.depth < layers.length
+        ? VerticalAlignments.bottom
+        : VerticalAlignments.top;
+      const fontSizes = allFontSizes[Math.min(node.depth, allFontSizes.length) - 1];
+      const {
+        textColor,
+        textInvertible,
+        fontStyle,
+        fontVariant,
+        fontFamily,
+        fontWeight,
+        valueFormatter,
+        padding,
+      } = Object.assign(
```

Comment:
can you add a type here that this merge is going to resolve to?

```suggestion
      }: MyType = Object.assign(
```

Reply:
There's currently no specific type for this, and making one up for this wouldn't be terribly informative as it's just the reiteration of what's in the `const` bracket. I'm planning further extraction from this function (not in this PR though) as it's still too large, in which case a proper boundary, carrying these resolved/cascaded values, needs to arise. But it needs a bit of design work, glad to chat about it on a call separately or on the team sync, the upshot is that I feel uncomfortable typing this transitional thing without that design. Still open to being convinced if I can learn about the reasons 👍

Reply:
Sure that's fine, this merging of several values is just hard to decipher. Even if the types were defined inline that would be better, something like...

```suggestion
      }: {
        fontWeight: number,
        valueFormatter: () => string,
        padding: number,
      } = Object.assign(
```

Reply:
Or even a combination of types that satisfy the resolved values, surely you have all the types to combine into the correct values.

```suggestion
      }: MyFontType & Pick<MyStyleType, 'style'> = Object.assign(
```